#### Prescription Drug Plan Formulary and Pharmacy Network Data Exploration

The Monthly Prescription Drug Plan Formulary and Pharmacy Network Information dataset is used in this analysis. The [dataset](https://data.cms.gov/provider-summary-by-type-of-service/medicare-part-d-prescribers/monthly-prescription-drug-plan-formulary-and-pharmacy-network-information) can be downloaded from Centers for [Medicare & Medicaid Services (CMS)](https://data.cms.gov/) website. The dataset is updated monthly. I use data for June 30, 2025 for the analysis.

This dataset contains detailed information about prescription drug plans (PDPs) and Medicare Advantage plans with drug coverage (MA-PDs). It includes plan identifiers, formulary IDs, premiums, deductibles, contract and plan names, and geographic coverage (state and county codes). The dataset also lists which drugs are included or excluded from a plan’s formulary and provides details on preferred pharmacy networks. It is primarily used by policymakers, researchers, and analysts to evaluate plan availability, pricing, and coverage variations across regions. The data supports public transparency and helps inform healthcare access and cost analyses.

I am exploring the dataset using PostgreSQL in the code blocks below. The questions are generated using ChatGPT.

In [2]:
#Import libraries
import pandas as pd
from sqlalchemy import create_engine, text

# Create database connection
engine = create_engine('postgresql+psycopg2://tharinduabeysinghe:#####@localhost/Pharmacy')

# Run quey and load data to a dataframe
def execute_sql_query(sql):
    # Load data into a pandas DataFrame
    df = pd.DataFrame()
    with engine.connect() as conn:
        df = pd.read_sql_query(text(sql), conn)
    return df

In [ ]:
# Number of drug formularies in the basic_drug_formulary table
sql_query = """
SELECT COUNT(DISTINCT formulary_id)
FROM basic_drug_formulary;
"""

# Execute query
execute_sql_query(sql)

count
0    309

In [ ]:
# Grain of the Plan information table
sql_query = """
SELECT 
	contract_id,
	plan_id,
	county_code,
	COUNT(*) AS cnt
FROM public.plan_information
GROUP BY contract_id, plan_id, county_code
HAVING COUNT(*) > 1;
"""

# Execute query
execute_sql_query(sql_query)

In [ ]:
# Grain of the basic drug formulary table
sql_query = """
SELECT
  formulary_id,
  rxcui,
  COUNT(*) AS cnt
FROM basic_drug_formulary
GROUP BY formulary_id, rxcui
HAVING COUNT(*) > 1;
"""

# Execute query
execute_sql_query(sql)

In [ ]:
# Number of drug formularies in the plan_information table
sql_query = """
SELECT COUNT(DISTINCT formulary_id) 
FROM plan_information;
"""

# Execute query
execute_sql_query(sql)

count
0    384

In [ ]:
# Number of plans in the plan_information table
sql_query = """
SELECT COUNT(DISTINCT plan_id) 
FROM plan_information;
"""

# Execute query
execute_sql_query(sql)

count
0    509

In [ ]:
# Number of drugs in the basic_drug_formulary table
sql_query = """
SELECT COUNT(DISTINCT rxcui) 
FROM basic_drug_formulary;
"""

# Execute query
execute_sql_query(sql)

count
0   5955

In [ ]:
# Number of formularies in each plan
sql_query = """
SELECT plan_id,
   	   COUNT(formulary_id) AS formulary_cnt
FROM plan_information
GROUP BY plan_id
ORDER BY formulary_cnt DESC;
"""

# Execute query
execute_sql_query(sql)


plan_id  formulary_cnt
0        256              1
1        181              1
2        257              1
3        240              1
4        338              1
..       ...            ...
504        5           3128
505        4           4024
506        3           5616
507        2           6601
508        1          10962

[509 rows x 2 columns]

In a plan, 

- Maximum drug count: 10962

- Minimum drug count: 1


In [ ]:
# Number of drugs in each formulary
sql_query = """
SELECT formulary_id,
   	   COUNT(rxcui) AS drug_cnt
FROM basic_drug_formulary
GROUP BY formulary_id
ORDER BY drug_cnt;
"""

# Execute query
execute_sql_query(sql)

formulary_id  drug_cnt
0           25420      1989
1           25419      2833
2           25413      2866
3           25071      2951
4           25179      3011
..            ...       ...
304         25389      4022
305         25022      4070
306         25020      4418
307         25016      4418
308         25014      4418

[309 rows x 2 columns]

In a formulary 

- Minimum drug count: 1989

- Maximum drug count: 4418

In [ ]:
# Number of contracts in a plan
sql_query = """
SELECT plan_id,
   	   COUNT(contract_id) AS contract_cnt
FROM plan_information
GROUP BY plan_id
ORDER BY contract_cnt DESC;
"""

# Execute query
execute_sql_query(sql)

plan_id  contract_cnt
0          1         10962
1          2          6601
2          3          5616
3          4          4024
4          5          3128
..       ...           ...
504      550             1
505      537             1
506      256             1
507      502             1
508      181             1

[509 rows x 2 columns]

In a plan

- Minimum contracts count: 1

- Maximum contracts count: 10962

1. How many drugs does each plan covers?

In this dataset, rxcui is a standardized system to identify drugs across the United States. 

In [ ]:
sql_query = """
SELECT
    plan_id,
    COUNT(rxcui) AS Drug_count
FROM public.basic_drug_formulary f
JOIN public.plan_information p
    ON f.formulary_id = p.formulary_id
GROUP BY plan_id
ORDER BY Drug_count DESC;
"""

# Execute query
execute_sql_query(sql)

plan_id  drug_count
0          1    30860214
1          2    19256908
2          3    16874958
3          4    12741796
4          5     9314944
..       ...         ...
451      580        3272
452      478        3272
453      536        3272
454      537        3272
455      350        3027

[456 rows x 2 columns]